src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptor.cs

```diff
@@ -40,6 +40,12 @@ protected TagHelperDescriptor(string kind)
 
         public IReadOnlyDictionary<string, string> Metadata { get; protected set; }
 
+        // Hoisted / cached metadata
```

Comment:
Really happy you didn't throw this in the actual Metadata data bag because that would have gotten messy real quick 👏

Reply:
Hm. Can you explain why using the `Metadata` bag is not right here? Is it because those values get serialized/passed around as public properties in the descriptor?

What are some conventions around when to use the Metadata bag and when to not?

Reply:
> Is it because those values get serialized/passed around as public properties in the descriptor?

Yup exactly, they actually also contribute to the hash code when requested; meaning calculated data (which this is since it serves as a computed cache) should never be in metadata because it changes the immutability of the TagHelper interface and its serialized form/hashcode.

> What are some conventions around when to use the Metadata bag and when to not?

Honestly Metadata is a bit of a hack and serializes poorly. For the "best" end-to-end solution I'd say try to never use Metadata; although it's a more expensive dev cost approach to not use it due to the requirement to update construction sites / custom JSON converters.

Reply:
The other thing is that Metadata requires a (virtual) dictionary lookup. A non-virtual property access is going to be significantly faster than that.